Selector plugin for selecting the planner and the controller

padhupradheep opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3269) |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | Neobotix MPO-700 |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

Now we can select the controller and the planner plugins on the fly

## Description of documentation updates required from your changes

Maybe RViz plugins after all needs a documentation somewhere?

---

## Future work that may be required in bullet points

* Goal checkers and smoothers could be added to the selection list if needed


#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_bt_navigator/behavior_trees/navigate_to_pose_w_replanning_and_recovery.xml:
```diff
@@ -8,6 +8,8 @@
   <BehaviorTree ID="MainTree">
     <RecoveryNode number_of_retries="6" name="NavigateRecovery">
       <PipelineSequence name="NavigateWithReplanning">
+        <ControllerSelector selected_controller="{selected_controller}" default_controller="FollowPath2" topic_name="controller_selector"/>
```
"FollowPath2"?

I assume the BT changes here and in the other file are temporary / not complete. In general, I like the first-step of setting the selectors as you did in the 2 files. Set the defaults to `GridBased` `FollowPath`, then use those blackboard entries in the FollowPath / ComputePathToPose BT nodes! 

Just make sure to update all the relevant BTs

padhupradheep replied:
> Just make sure to update all the relevant BTs

You mean the docs right?

SteveMacenski commented on nav2_rviz_plugins/src/selector.cpp:
```diff
@@ -0,0 +1,129 @@
+void
+Selector::start_ros_timer()
+{
+  rclcpp::spin(node_);
```
Where is this `node_` set?

SteveMacenski reviewed (commented):
Generally looks good to me! What's this tool look like? :smile:

SteveMacenski commented:
Also, seems easy enough to add the smoother / goal checker / progress checkers here too! Nice, simple tool, I like it!

SteveMacenski commented:
Progress checker too :heart: 

I think then its just the BT updates and perhaps this is good for testing / merging? What do you think?

padhupradheep commented:
> Progress checker too

There is no BT node for that as far as I checked. We can use the dynamic reconf server for it but that would be something unique and really won't fit to the scope of this PR. 

> I think then its just the BT updates and perhaps this is good for testing / merging? What do you think?

Yes, almost done!  I need to update the docs and probably you can give it a swing when you get some time and we are all set for the merge once after that.

padhupradheep commented:
Shall I add this plugin as default in nav2 bringup ? 

Since it's more of a development tool, I feel it would be better if the user selects it by hand when required.

SteveMacenski commented:
Yeah that sounds great!

padhupradheep commented:
![image](https://github.com/ros-planning/navigation2/assets/20242192/5da0a74c-2df0-4a38-8874-d8d552e3f700)

bit more compact version..

Would this work?

SteveMacenski commented:
Yeah that new 2x2 version works, though why not have the drop down on the same line as the label instead of a new line? Either way works.

If you're going to make a new gif with the format, try to use not just `FollowPath` and `FollowPath2`, but how about `FollowPath` and `AvoidObstacles` or something else with different modality implications.

padhupradheep commented:
> why not have the drop down on the same line as the label instead of a new line

I tried it yesterday, It felt cramped by this way.

![image](https://github.com/ros-planning/navigation2/assets/20242192/4c6379ad-8c29-4a51-910d-25119ad27b69)

I'd stick to the other version..

padhupradheep commented:
Just leaving a short note on having the selector plugin has default in the rviz bringup. There is some sort of race condition that does not allow the selector plugin to populate atleast one of the component's combo boxes, atleast once in ten instances.

SteveMacenski commented:
> There is some sort of race condition that does not allow the selector plugin to populate atleast one of the component's combo boxes, atleast once in ten instances.

Do you see your error log about the server not being up yet? Can you change the ` auto parameters = parameter_client->get_parameters({plugin_type});` line to use the API which would give more feedback on failure?

It would be good to know where this is happening so we can see what might need to be retried / delayed to get it to work deterministically. 1/10 failures is not something we could merge :(

padhupradheep commented on nav2_rviz_plugins/src/selector.cpp:
```diff
@@ -0,0 +1,184 @@
+
+  // Wait for the service to be available before calling it
+  bool server_unavailable = false;
+  while (!parameter_client->wait_for_service(3s)) {
```
We get the issue exactly here. As you see indeed have a logger for it too below. It's always the `controller_server` which is not available.. Probably we could wait couple of more seconds here?

SteveMacenski replied:
Possibly yes! Or you could wait once, then just pass the callback and come back again in another 200ms to try again, rather than `while` looping

padhupradheep commented:
Issue is fixed! Thanks for the suggestion! :+1:

SteveMacenski commented:
CI had error I assume isn't related to this PR, but retriggering to check.

In the meantime, anything else you want to do before I test and merge?

padhupradheep commented:
> In the meantime, anything else you want to do before I test and merge?

Well, I just added the progress_checker too. I just updated from 2x2 to 3x2 configuration, with slight changes to the way combo boxes would be placed. 

Now I think, it's ready to be tested.

SteveMacenski commented:
> CI had error I assume isn't related to this PR, but retriggering to check.

Huh, your PR is regularly triggering these failures. Can you reproduce locally? Maybe try pulling in `main`?

padhupradheep commented:
I already did a rebase early today.. not sure what's causing the issue. Let me check..

SteveMacenski commented:
Beyond that, I tested it locally and looks good to me

padhupradheep commented:
For some reason this is happening..

` [bt_navigator-14] [ERROR] [1707415321.396377533] [bt_navigator]: Exception when loading BT: Error at line 10: -> Node not recognized: ControllerSelector`

SteveMacenski commented:
>  [bt_navigator-14] [ERROR] [1707415321.396377533] [bt_navigator]: Exception when loading BT: Error at line 10: -> Node not recognized: ControllerSelector

Oh. Note that you have to update the BT nodes to default load in a few places (also, add in progress checker while you're at it, I totally missed it too): 

- https://github.com/ros-planning/navigation2/blob/main/nav2_bt_navigator/src/bt_navigator.cpp
- https://github.com/ros-planning/navigation2/blob/main/nav2_bringup/params/nav2_params.yaml (and other 3 next to it)

Crucially the following:
- https://github.com/ros-planning/navigation2/blob/0cdabca2f27191da7ef3f8f1edff75e9e2e925aa/nav2_system_tests/src/costmap_filters/keepout_params.yaml#L50
- https://github.com/ros-planning/navigation2/blob/0cdabca2f27191da7ef3f8f1edff75e9e2e925aa/nav2_system_tests/src/costmap_filters/speed_local_params.yaml#L50
- https://github.com/ros-planning/navigation2/blob/0cdabca2f27191da7ef3f8f1edff75e9e2e925aa/nav2_system_tests/src/gps_navigation/nav2_no_map_params.yaml#L18
- https://github.com/ros-planning/navigation2/blob/0cdabca2f27191da7ef3f8f1edff75e9e2e925aa/nav2_system_tests/src/costmap_filters/speed_global_params.yaml#L50

This requirement is not long for this world, @facontidavide opened a PR which has a tool which autopopulates this which will remove this need in the coming weeks. But for now, we manually set the BT nodes in the plugins lib list.

padhupradheep commented:
Ohhh.. this again.. I remember we faced the same trouble when we developed all the `cancel BTs` last year.

Glad that it's not needed anymore..

padhupradheep commented:
Okay, now the release_test fails because of someother reason.. :thinking:

SteveMacenski commented:
What do you mean? Its failing for the exact same reason: https://app.circleci.com/pipelines/github/ros-planning/navigation2/10965/workflows/724ba00e-3d4b-49ec-a289-d41bf0a1e4ac/jobs/34394/tests

It doesn't seem that you pushed any updates to add the new BT nodes like we discussed.

padhupradheep commented:
My mistake, somehow messed up the commit while fixing the DCO. I just was following the recommendation it was giving out, but somehow it broke my changes :man_shrugging:

padhupradheep commented:
https://app.circleci.com/pipelines/github/ros-planning/navigation2/10966/workflows/2b3de681-a592-4539-866f-bcfe9e88dedc/jobs/34400/tests

Now it seems to be the `test_spin_behavior_fake.missing_result` and  `test_drive_on_heading_recovery.missing_result`

SteveMacenski commented:
I was messing around there recently, but it all passed on my CI jobs. Maybe increment your cache to v20 in the 3x places in the circleci file to break it all? If that still fails, I'll take that as an action item to look into, I won't block you further with it.

padhupradheep commented:
v20 didn't help either..

please let me know, if there is something else that I can try to sort it outt.

SteveMacenski commented:
Not your problem, probably me. If you wanted to take a look at it, the more the merrier :wink: 

Want to post on discourse with your new feature + gif? I think that's worthy of highlighting!

padhupradheep commented:
> Not your problem, probably me. If you wanted to take a look at it, the more the merrier 😉

:smile: 

> Want to post on discourse with your new feature + gif? I think that's worthy of highlighting!

Oh! Sure can do that.. :+1:

padhupradheep commented:
It took me one whole year to eventually get on this PR and get it up.. :face_with_spiral_eyes: 

I'll clear #3347 too..

If you have anything else, feel free to tag me in..

SteveMacenski commented:
Thanks and will do!
